Using Vert.x guide

tsegismont opened the pull request:
Guide for the Vert.x extension

tsegismont committed:
Reactive web resources documentation

Includes Vert.x extension installation, async and streaming over SSE.

tsegismont committed:
Vert.x JSON as web resource result

tsegismont committed:
Lightweight messaging with the EventBus

Also introduces the Axle shim

cescoffier committed:
Start restructuring the "using vert.x" doc

cescoffier committed:
Write the async message passing guide

cescoffier committed:
Restructured the using-vertx doc

tsegismont commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,239 @@
+mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
+    -DprojectGroupId=org.acme \
+    -DprojectArtifactId=vertx-quickstart \
+    -Dextensions="vert.x"
```
```suggestion
    -Dextensions="vertx"
```

cescoffier replied:
Both are supported, but yes, `vertx` is one less keystroke.

tsegismont commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,334 @@
+mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
+    -DprojectGroupId=org.acme \
+    -DprojectArtifactId=vertx-quickstart \
+    -Dextensions="vert.x"
```
```suggestion
    -Dextensions="vertx"
```

tsegismont commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,334 @@
+If you are familar with Vert.x, you know that Vert.x provides different API models.
+For instance _bare_ Vert.x uses callbacks, the RX Java 2 version uses `Single`, `Maybe`, `Completable`, `Observable` and `Flowable`.
+
+{project-name} provides 3 Vert.x APIs:
```
My initial commit had a paragraph explaining why users would choose one or the others. Why not keeping it? I'm afraid it would be confusing to say there are 3 different APIs without explaining the rationale.

cescoffier replied:
I've re-added the paragraph.

tsegismont commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,334 @@
+
+[source, shell]
+----
+Hello Quarkus! (10 ms)
```
Don't forget to update the `quickstarts` tests too.

tsegismont commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,334 @@
+[source, java]
+----
+// Use io.vertx.axle.core.Vertx;
+@Inject Vertx vertx;
```
For this example it would be better to do the tick stream transformations with Reactive MP operators, I think. After all, the Axle shim is designed to work well with Quarkus and MP reactive APIs.

cescoffier replied:
Very good point, it's back!

tsegismont commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,334 @@
+
+[source, text]
+----
+{"Hello":"Quarkus"}
```
Same here, don't forget to update the `quickstarts`.

tsegismont commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,334 @@
+
+Needless to say, this works equally well when the JSON result is wrapped in a `CompletionStage` or a `Publisher`.
+
+== Using the event bus
```
How about renaming this section to "Going further", or "Beyond the basics"... whatever.

Then below we can link to different quides:
- async message passing (in the making)
- web client (in the making)
- pg client
- ...

cescoffier replied:
I was thinking about having a whole section about clients. We also need to cover the `@Route`.

tsegismont replied:
Not sure what you mean? You convinced me it was a good idea to split the content. Where do you want to add the client section?

cescoffier replied:
I've fixed this, and you are right it's already long enough. I've added a small section about using clients.

tsegismont reviewed (commented):
@cescoffier some comments about your changes

tsegismont commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,239 @@
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class NameService {
```
Don't forget to update the `quickstarts``

tsegismont commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,239 @@
+public class NameService {
+
+    @ConsumeEvent                           // <1>
+    public String consume(String name) {    // <2>
```
I think a warning should be added indicating the message will be processed on an event loop, and the method should not use blocking APIs.

tsegismont replied:
Or we simply delay the documentation and revisit the implementation so that events are consumed on the regular xnio/undertows tasks thread pool.

cescoffier replied:
We are adding a `blocking` attribute (pending PR)

cescoffier replied:
But for now (as I don't want to wait for the release to publish these docs), I agree it must be mentioned.

cescoffier replied:
Done!

tsegismont commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,239 @@
+}
+----
+
+This bean receives the name, and return the greeting message.
```
Originally, I added this content about async message passing as a part of a single _Going Reactive with Vert.x_ document. The other document now has a section about compiling and starting the dev mode, but not this one.

It would be nice to put it here again, I think, along with a link to try so that someone who has started the `quickstarts` in the background can see the results immediately.

cescoffier replied:
I've extended the guide with this. 
I've also pushed changes to the quickstart.

tsegismont reviewed (commented):
@cescoffier another pass on the async message passing document

cescoffier committed:
Consistent usage of "vertx" to add the extension

cescoffier committed:
Add a paragraph explaining why there are 3 APIs.

This paragraph was there initially (in Thomas commits), moved, deleted, but finally back...

cescoffier committed:
Add a note about the non-blocking aspect of the code processing messages.

cescoffier commented:
@tsegismont I will start working on the quickstart part Today (that's why it was not updated or was missing in the guide).

cescoffier committed:
Sync up with the quickstart content

cescoffier commented:
Now the question is: what else do we want absolutely for the first version.

cescoffier committed:
Add section about Vert.x client and update the conclusion

cescoffier committed:
Allow vert.x and axle variables to be injected in the documentation

tsegismont commented:
@cescoffier thanks for adding the web client sample. I believe we're good as a initial version. The section on configuration can be contributed later.

cescoffier commented:
@tsegismont I agree about the configuration. I still need to work on it (not on the doc, on the configuration itself ;-)). We had "recent" changes around configuration and it deserves a bit more testing.

cescoffier commented:
@FroMage as you are back, can you have a look? As I worked on it directly, I cannot really "approve" it.

cescoffier commented:
@FroMage Up?

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+<2> The method parameter is the message body. If the method returns _something_ it's the message response.
+
+// TODO Revisit this once the blocking PR is integrated - https://github.com/quarkusio/quarkus/pull/1646
+IMPORTANT: the code consuming the event must be _non-blocking_, as it's called on the Vert.x event loop.
```
s/the/The

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+    return name.toUpperCase();
+}
+----
+<1> Receive the messages sent on the `greeting` address
```
isn't it "sent to"?

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+
+=== Replying
+
+The _return_ of a method annotated with `@ConsumeEvent` is used as response to the incoming message.
```
You first give examples with returning a value, then do "consuming without replying" then document replying. I think you should have started with not having a reply and then documenting how to reply later, but not back and forth.

cescoffier replied:
This choice was made to be in sync with the quickstart. Trying to see if we can change it without having to change too much in the quickstart.

FroMage replied:
Just don't change the quickstart ;)

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+
+=== Replying
+
+The _return_ of a method annotated with `@ConsumeEvent` is used as response to the incoming message.
```
s/return/return value/

FroMage replied:
This was not fixed :(

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+----
+@ConsumeEvent("greeting")
+public CompletionStage<String> consume2(String name) {
+    return CompletableFuture.supplyAsync(name::toUpperCase);
```
Oh sure, great, start telling people to run shit in the JDK's default executor thread pool :(

cescoffier replied:
Changed

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+
+  @GET
+  @Path("/{name}")
+  public CompletionStage<String> hello(@PathParam("name") String name) {
```
RESTEasy has a version of `@PathParam` that does not require duplicating the parameter name. Probably we want to use this.

cescoffier replied:
Didn't know this!

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+
+    @GET
+    @Path("/async/{name}")
+    public CompletionStage<String> hello(@PathParam("name") String name) {
```
Same remark about `@PathParam`

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+}
+----
+
+This bean receives the name, and return the greeting message.
```
s/return/returns/

FroMage commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+To better understand, let's detail how the HTTP request/response has been handled:
+
+1. The request is received by the `hello` method
+2. a message containing the _name_ is sent on the event bus
```
s/a/A/

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+Eclipse https://vertx.io[Vert.x] is a toolkit for building reactive applications.
+It is designed to be lightweight and embeddable.
+Vert.x defines a reactive execution model and provides a large ecosystem.
+Quarkus integrates Vert.x to implement different forms of reactive.
```
Oh yeah? "different forms of reactive"? Any specific reactive thing or all things reactive? I think there's a word missing ;)

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+Quarkus integrates Vert.x to implement different forms of reactive.
+Basically, {project-name} uses Vert.x as its reactive engine.
+While lots of reactive features from {project-name} don't _show_ Vert.x, it's used underneath.
+But you can also access the managed Vert.x instance and benefits from the Vert.x ecosystem.
```
s/benefits/benefit/

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+
+== Accessing Vert.x
+
+Once the extension added, you can access the _managed_ Vert.x instance using `@Inject`:
```
s/extension added/extension has been added/

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+
+// Axle API:
+vertx.fileSystem().readFile("lorem-ipsum.txt")
+    .thenApply(buffer -> "Content: " + buffer.toString("UTF-8"))
```
You're cheating here: it does not implement things like the Rx version above, they should be much more similar.

cescoffier replied:
Replaced with `whenComplete`. However, each model comes with its own idioms and I generally change a bit the processing depending on the API I have.

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+    .thenAccept(s -> System.out.println(s));
+----
+
+== Using Vert.x in Reactive Jax-RS resources
```
s/Jax-RS/JAX-RS/

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+    String message = String.format("Hello %s! (%d ms)%n", name, duration);
+
+    // Complete
+    future.complete(message);
```
Doesn't the vert axle flavour have a version of `setTimer` that returns a `CompletionStage` we can use to simplify this example?

cescoffier replied:
That's a good idea, but no, it does not provide this method at the moment. 
The only difference with the bare vert.x version is the usage of `Consumer` instead of `Handler`

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+{project-name} web resources that need to send content as https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events[server-sent events] must have a method:
+
+* declaring the `text/event-stream` response content type
+* returning a https://www.reactive-streams.org/[Reactive Streams] `Publisher`
```
Or an RxJava 1&2 Observable (but only Rx2 is relevant to this guide since you mention it earlier)

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+@Produces(MediaType.SERVER_SENT_EVENTS)
+@Path("{name}/streaming")
+public Publisher<String> greeting(@PathParam("name") String name) {
+    return ReactiveStreams.fromPublisher(vertx.periodicStream(2000).toPublisher())
```
@cescoffier do we have a good reason why the axle flavours can't have a `toReactiveStreams()` method that does `ReactiveStreams.fromPublisher(this.toPublisher())`?

cescoffier replied:
It's an opened issue on Axle. It should be `.toPublisherBuilder()`. 
https://github.com/smallrye/smallrye-reactive-utils/issues/15

FroMage replied:
OK cool.

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+public Publisher<String> greeting(@PathParam("name") String name) {
+    return ReactiveStreams.fromPublisher(vertx.periodicStream(2000).toPublisher())
+            .map(l -> String.format("Hello %s! (%s)%n", name, new Date()))
+            .buildRs();
```
@cescoffier I hope you have a good reason to have named this method `buildRs()` and not `toPublisher()` because from a user PoV I just did `toPublisher`, `fromPublisher` and then `buildRs` sticks out like a bloody thorn, especially when I want to return a damn `publisher`

cescoffier replied:
This comes from the spec. In Java 9 it will be `build` (and use Flow).
I hope it would be possible to return a `PublisherBuilder` with the right _provider_.

FroMage replied:
Well, the spec was crazy to allow a method name with an abbreviation like `buildRs`, sorry :(

cescoffier replied:
It's the JDK8/JDK9 issue... because we had to plan backward compatibility we cannot use `build()`:

```
/*ReactiveStream*/ Publisher<X> build();
/*Flow*/ Publisher<X> build();
```

While I agree it's not great, it's going to be hidden as soon as Flow becomes the main model.

FroMage replied:
Well, `buildReactiveStream()` was still available ;)

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+  <version>{axle-version}</version>
+</dependency>
+
+<!-- RX Java 2 API -->
```
I think you should add a few words to explain why rxjava2 does not come from a `vertx-rx-java2-web-client` module.

cescoffier replied:
Good point.

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+        </goals>
+        <configuration>
+            <enableHttpUrlHandler>true</enableHttpUrlHandler>
+            <enableHttpsUrlHandler>true</enableHttpsUrlHandler> <!-- Add this line -->
```
Aren't those set automatically with an ssl config option?

cescoffier replied:
Good point.

cescoffier replied:
This line set native ssl to `true`, I don't know if the opposite it true, let me check.

cescoffier replied:
So, both work :-D 

I've changed to use the `application.properties`

FroMage commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -0,0 +1,483 @@
+mvn package -Pnative
+----
+
+///// TODO - Configuring Vert.x
```
Is this normal?

cescoffier replied:
Let me remove it.

FroMage reviewed (changes requested):
THIS IS WHAT YOU GET IF YOU ASK ME TO REVIEW TWICE ;)

cescoffier commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+@ApplicationScoped
+public class GreetingService {
+
+    @ConsumeEvent                           // <1>
```
`@ConsumeEvent` uses the Vert.x event bus, it's somehow related to reactive messaging but not directly. 

`@ConsumeEvent("my-eventbus-address")` is equivalent to `@Incoming("my-stream")` configured as follows:

```
mp.messaging.incoming.my-stream.type=...VertxEventBus
mp.messaging.incoming.my-stream.address="my-eventbus-address"
```

However, where it's different is about the _reply-ability_, `@ConsumeEvent` can reply to the message, `@Incoming` does not.

FroMage replied:
I'm only asking because if we have both types of messaging we're going to have to explain why and which one to use in which cases.

cescoffier replied:
They have very different intents, but we may need to consider 3 of them:

* Reactive messaging is about streaming and "distributed" messages.
* `@ConsumeEvent` is about local events only.
* CDI events

The CDI events semantic is very different from the event bus (no reply, no 1 -> 1). I'm more puzzled about having CDI events and `@ConsumeEvent`. 

One thing not done yet is to completely hide the Vert.x event bus on the sender side.

FroMage replied:
OK, then just explain this in the doc and we're good.

cescoffier replied:
I've added an admonition.

cescoffier commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,290 @@
+    }
+}
+----
+<1> If not set, the address is the fully qualified name of the bean, for instance, in this snippet it's `org.acme.NameService`
```
Yep, good catch.

cescoffier replied:
package name was also wrong.

cescoffier committed:
Apply the review from Stephane Epardaud

cescoffier committed:
Apply second round of reviews

cescoffier commented:
I've done a second round on the reviews. Hopefully, everything has been addressed.

mkouba commented on docs/src/main/asciidoc/async-message-passing.adoc:
```diff
@@ -0,0 +1,295 @@
+<1> If not set, the address is the fully qualified name of the bean, for instance, in this snippet it's `org.acme.vertx.GreetingService`.
+<2> The method parameter is the message body. If the method returns _something_ it's the message response.
+
+// TODO Revisit this once the blocking PR is integrated - https://github.com/quarkusio/quarkus/pull/1646
```
FYI https://github.com/quarkusio/quarkus/pull/1646 was already merged.

cescoffier replied:
Yes, but we want to publish this guide before the actual release of Quarkus. Quarkus release may be delayed due to a few regressions.

mkouba replied:
Np, I'm just saying it's there ;-).

FroMage commented:
Almost, there's one grammar fix left.

cescoffier committed:
fix missing word in documentation

cescoffier commented:
@FroMage Damn! Fixed.

I don't understand why it was not displayed initially.
